lib/rubocop/cop/lint/mutable_literal.rb

```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Lint
+      # Guards against bugs from mutable literals used as default arguments during
+      # Hash initialization. Since that will return the same mutable object
+      # for all missing keys, it would allow each mutation to affect the values of all keys
+      # initialized with the default value simultaneously.
+      #
+      # @example
+      #   # bad
+      #   Hash.new([])
+      #   Hash.new({})
+      #   Hash.new(Array.new)
+      #   Hash.new(Hash.new)
+      #
+      #   # good
```

Comment:
Thanks, I was coming at this from the perspective that "Never mutate shared state" is the authoritative fix for this problem.  i.e. use `+=` rather than `<<` (unless you are in a performance critical path and then make certain you are doing exactly and only what's needed). 

From that point of view, this rule simply prevents others from impinging on the state of those doing it correctly.

```
h = Hash.new { [] }
pp h       # => {}
pp h[:foo] # => []
h[:foo] += [123]
pp h       # => {:foo=>[123]}
pp h[:foo] # => [123]
```

In that sense, this approach would be the lighter touch version of `.freeze` since it does not error.

However, given how broadly this will be applied, I agree we don't want anyone falling into a "lost mutations" trap either.

I re-labeled these as `@bad` but out of scope

Reply:
Also relevant: 

* https://github.com/rails/rails/issues/3184#issuecomment-2257200

Reply:
Okay, I went hard on describing the examples, I actually like it a lot more now. 

The options and choices are much clearer.